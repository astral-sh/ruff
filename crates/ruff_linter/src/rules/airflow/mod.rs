//! Airflow-specific rules.
pub(crate) mod rules;

#[cfg(test)]
mod tests {
    use std::path::Path;

    use anyhow::Result;
    use test_case::test_case;

    use crate::registry::Rule;
    use crate::test::test_path;
    use crate::{assert_messages, settings};

    #[test_case(Rule::AirflowVariableNameTaskIdMismatch, Path::new("AIR001.py"))]
    #[test_case(Rule::AirflowDagNoScheduleArgument, Path::new("AIR301.py"))]
    #[test_case(Rule::Airflow3Removal, Path::new("AIR302_args.py"))]
    #[test_case(Rule::Airflow3Removal, Path::new("AIR302_names.py"))]
    #[test_case(Rule::Airflow3MovedToProvider, Path::new("AIR303.py"))]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.noqa_code(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("airflow").join(path).as_path(),
            &settings::LinterSettings::for_rule(rule_code),
        )?;
        assert_messages!(snapshot, diagnostics);
        Ok(())
    }
}
