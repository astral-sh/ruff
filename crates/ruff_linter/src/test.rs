#![cfg(any(test, fuzzing))]
//! Helper functions for the tests of rule implementations.

use std::borrow::Cow;
use std::path::Path;

#[cfg(not(fuzzing))]
use anyhow::Result;
use itertools::Itertools;
use rustc_hash::FxHashMap;

use ruff_diagnostics::{Diagnostic, FixAvailability};
use ruff_python_ast::PySourceType;
use ruff_python_codegen::Stylist;
use ruff_python_index::Indexer;
use ruff_python_parser::lexer::LexResult;
use ruff_python_parser::AsMode;
use ruff_python_trivia::textwrap::dedent;
use ruff_source_file::{Locator, SourceFileBuilder};
use ruff_text_size::Ranged;

use crate::directives;
use crate::fix::{fix_file, FixResult};
use crate::linter::{check_path, LinterResult};
use crate::message::{Emitter, EmitterContext, Message, TextEmitter};
use crate::packaging::detect_package_root;
use crate::registry::AsRule;
use crate::rules::pycodestyle::rules::syntax_error;
use crate::settings::types::UnsafeFixes;
use crate::settings::{flags, LinterSettings};
use crate::source_kind::SourceKind;
use ruff_notebook::Notebook;
#[cfg(not(fuzzing))]
use ruff_notebook::NotebookError;

#[cfg(not(fuzzing))]
pub(crate) fn test_resource_path(path: impl AsRef<Path>) -> std::path::PathBuf {
    Path::new("./resources/test/").join(path)
}

/// Run [`check_path`] on a file in the `resources/test/fixtures` directory.
#[cfg(not(fuzzing))]
pub(crate) fn test_path(path: impl AsRef<Path>, settings: &LinterSettings) -> Result<Vec<Message>> {
    let path = test_resource_path("fixtures").join(path);
    let contents = std::fs::read_to_string(&path)?;
    Ok(test_contents(&SourceKind::Python(contents), &path, settings).0)
}

#[cfg(not(fuzzing))]
pub(crate) struct TestedNotebook {
    pub(crate) messages: Vec<Message>,
    pub(crate) source_notebook: Notebook,
    pub(crate) linted_notebook: Notebook,
}

#[cfg(not(fuzzing))]
pub(crate) fn test_notebook_path(
    path: impl AsRef<Path>,
    expected: impl AsRef<Path>,
    settings: &LinterSettings,
) -> Result<TestedNotebook, NotebookError> {
    let source_notebook = Notebook::from_path(path.as_ref())?;

    let source_kind = SourceKind::IpyNotebook(source_notebook);
    let (messages, transformed) = test_contents(&source_kind, path.as_ref(), settings);
    let expected_notebook = Notebook::from_path(expected.as_ref())?;
    let linted_notebook = transformed.into_owned().expect_ipy_notebook();

    assert_eq!(
        linted_notebook.cell_offsets(),
        expected_notebook.cell_offsets()
    );
    assert_eq!(linted_notebook.index(), expected_notebook.index());
    assert_eq!(
        linted_notebook.source_code(),
        expected_notebook.source_code()
    );

    Ok(TestedNotebook {
        messages,
        source_notebook: source_kind.expect_ipy_notebook(),
        linted_notebook,
    })
}

/// Run [`check_path`] on a snippet of Python code.
pub fn test_snippet(contents: &str, settings: &LinterSettings) -> Vec<Message> {
    let path = Path::new("<filename>");
    let contents = dedent(contents);
    test_contents(&SourceKind::Python(contents.into_owned()), path, settings).0
}

thread_local! {
    static MAX_ITERATIONS: std::cell::Cell<usize> = std::cell::Cell::new(8);
}

pub fn set_max_iterations(max: usize) {
    MAX_ITERATIONS.with(|iterations| iterations.set(max));
}

pub(crate) fn max_iterations() -> usize {
    MAX_ITERATIONS.with(std::cell::Cell::get)
}

/// A convenient wrapper around [`check_path`], that additionally
/// asserts that fixes converge after a fixed number of iterations.
pub(crate) fn test_contents<'a>(
    source_kind: &'a SourceKind,
    path: &Path,
    settings: &LinterSettings,
) -> (Vec<Message>, Cow<'a, SourceKind>) {
    let source_type = PySourceType::from(path);
    let tokens: Vec<LexResult> =
        ruff_python_parser::tokenize(source_kind.source_code(), source_type.as_mode());
    let locator = Locator::new(source_kind.source_code());
    let stylist = Stylist::from_tokens(&tokens, &locator);
    let indexer = Indexer::from_tokens(&tokens, &locator);
    let directives = directives::extract_directives(
        &tokens,
        directives::Flags::from_settings(settings),
        &locator,
        &indexer,
    );
    let LinterResult {
        data: (diagnostics, _imports),
        error,
    } = check_path(
        path,
        path.parent()
            .and_then(|parent| detect_package_root(parent, &settings.namespace_packages)),
        tokens,
        &locator,
        &stylist,
        &indexer,
        &directives,
        settings,
        flags::Noqa::Enabled,
        source_kind,
        source_type,
    );

    let source_has_errors = error.is_some();

    // Detect fixes that don't converge after multiple iterations.
    let mut iterations = 0;

    let mut transformed = Cow::Borrowed(source_kind);

    if diagnostics
        .iter()
        .any(|diagnostic| diagnostic.fix.is_some())
    {
        let mut diagnostics = diagnostics.clone();

        while let Some(FixResult {
            code: fixed_contents,
            source_map,
            ..
        }) = fix_file(
            &diagnostics,
            &Locator::new(transformed.source_code()),
            UnsafeFixes::Enabled,
        ) {
            if iterations < max_iterations() {
                iterations += 1;
            } else {
                let output = print_diagnostics(diagnostics, path, &transformed);

                panic!(
                    "Failed to converge after {} iterations. This likely \
                     indicates a bug in the implementation of the fix. Last diagnostics:\n{}",
                    max_iterations(),
                    output
                );
            }

            transformed = Cow::Owned(transformed.updated(fixed_contents, &source_map));

            let tokens: Vec<LexResult> =
                ruff_python_parser::tokenize(transformed.source_code(), source_type.as_mode());
            let locator = Locator::new(transformed.source_code());
            let stylist = Stylist::from_tokens(&tokens, &locator);
            let indexer = Indexer::from_tokens(&tokens, &locator);
            let directives = directives::extract_directives(
                &tokens,
                directives::Flags::from_settings(settings),
                &locator,
                &indexer,
            );

            let LinterResult {
                data: (fixed_diagnostics, _),
                error: fixed_error,
            } = check_path(
                path,
                None,
                tokens,
                &locator,
                &stylist,
                &indexer,
                &directives,
                settings,
                flags::Noqa::Enabled,
                &transformed,
                source_type,
            );

            if let Some(fixed_error) = fixed_error {
                if !source_has_errors {
                    // Previous fix introduced a syntax error, abort
                    let fixes = print_diagnostics(diagnostics, path, source_kind);

                    let mut syntax_diagnostics = Vec::new();
                    syntax_error(&mut syntax_diagnostics, &fixed_error, &locator);
                    let syntax_errors = print_diagnostics(syntax_diagnostics, path, &transformed);

                    panic!(
                        r#"Fixed source has a syntax error where the source document does not. This is a bug in one of the generated fixes:
{syntax_errors}
Last generated fixes:
{fixes}
Source with applied fixes:
{}"#,
                        transformed.source_code()
                    );
                }
            }

            diagnostics = fixed_diagnostics;
        }
    }

    let source_code = SourceFileBuilder::new(
        path.file_name().unwrap().to_string_lossy().as_ref(),
        source_kind.source_code(),
    )
    .finish();

    let messages = diagnostics
        .into_iter()
        .map(|diagnostic| {
            let rule = diagnostic.kind.rule();
            let fixable = diagnostic.fix.is_some();

            match (fixable, rule.fixable()) {
                (true, FixAvailability::Sometimes | FixAvailability::Always)
                | (false, FixAvailability::None | FixAvailability::Sometimes) => {
                    // Ok
                }
                (true, FixAvailability::None) => {
                    panic!("Rule {rule:?} is marked as non-fixable but it created a fix. Change the `Violation::FIX_AVAILABILITY` to either `FixAvailability::Sometimes` or `FixAvailability::Always`");
                },
                (false, FixAvailability::Always) => {
                    panic!("Rule {rule:?} is marked to always-fixable but the diagnostic has no fix. Either ensure you always emit a fix or change `Violation::FIX_AVAILABILITY` to either `FixAvailability::Sometimes` or `FixAvailability::None")
                }
            }

            assert!(!(fixable && diagnostic.kind.suggestion.is_none()), "Diagnostic emitted by {rule:?} is fixable but `Violation::fix_title` returns `None`.`");

            // Not strictly necessary but adds some coverage for this code path
            let noqa = directives.noqa_line_for.resolve(diagnostic.start());

            Message::from_diagnostic(diagnostic, source_code.clone(), noqa)
        })
        .sorted()
        .collect();
    (messages, transformed)
}

fn print_diagnostics(diagnostics: Vec<Diagnostic>, path: &Path, source: &SourceKind) -> String {
    let filename = path.file_name().unwrap().to_string_lossy();
    let source_file = SourceFileBuilder::new(filename.as_ref(), source.source_code()).finish();

    let messages: Vec<_> = diagnostics
        .into_iter()
        .map(|diagnostic| {
            let noqa_start = diagnostic.start();

            Message::from_diagnostic(diagnostic, source_file.clone(), noqa_start)
        })
        .collect();

    if let Some(notebook) = source.as_ipy_notebook() {
        print_jupyter_messages(&messages, path, notebook)
    } else {
        print_messages(&messages)
    }
}

pub(crate) fn print_jupyter_messages(
    messages: &[Message],
    path: &Path,
    notebook: &Notebook,
) -> String {
    let mut output = Vec::new();

    TextEmitter::default()
        .with_show_fix_status(true)
        .with_show_fix_diff(true)
        .with_show_source(true)
        .with_unsafe_fixes(UnsafeFixes::Enabled)
        .emit(
            &mut output,
            messages,
            &EmitterContext::new(&FxHashMap::from_iter([(
                path.file_name().unwrap().to_string_lossy().to_string(),
                notebook.index().clone(),
            )])),
        )
        .unwrap();

    String::from_utf8(output).unwrap()
}

pub(crate) fn print_messages(messages: &[Message]) -> String {
    let mut output = Vec::new();

    TextEmitter::default()
        .with_show_fix_status(true)
        .with_show_fix_diff(true)
        .with_show_source(true)
        .with_unsafe_fixes(UnsafeFixes::Enabled)
        .emit(
            &mut output,
            messages,
            &EmitterContext::new(&FxHashMap::default()),
        )
        .unwrap();

    String::from_utf8(output).unwrap()
}

#[macro_export]
macro_rules! assert_messages {
    ($value:expr, $path:expr, $notebook:expr) => {{
        insta::with_settings!({ omit_expression => true }, {
            insta::assert_snapshot!(
                $crate::test::print_jupyter_messages(&$value, &$path, &$notebook)
            );
        });
    }};
    ($value:expr, @$snapshot:literal) => {{
        insta::with_settings!({ omit_expression => true }, {
            insta::assert_snapshot!($crate::test::print_messages(&$value), $snapshot);
        });
    }};
    ($name:expr, $value:expr) => {{
        insta::with_settings!({ omit_expression => true }, {
            insta::assert_snapshot!($name, $crate::test::print_messages(&$value));
        });
    }};
    ($value:expr) => {{
        insta::with_settings!({ omit_expression => true }, {
            insta::assert_snapshot!($crate::test::print_messages(&$value));
        });
    }};
}
