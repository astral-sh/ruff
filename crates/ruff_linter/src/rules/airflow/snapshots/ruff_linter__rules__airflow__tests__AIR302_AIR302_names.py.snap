---
source: crates/ruff_linter/src/rules/airflow/mod.rs
snapshot_kind: text
---
AIR302_names.py:52:1: AIR302 `airflow.PY36` is removed in Airflow 3.0
   |
50 | from airflow.www.utils import get_sensitive_variables_fields, should_hide_value_for_key
51 | 
52 | PY36, PY37, PY38, PY39, PY310, PY311, PY312
   | ^^^^ AIR302
53 | 
54 | AWSAthenaHook
   |
   = help: Use `sys.version_info` instead

AIR302_names.py:52:7: AIR302 `airflow.PY37` is removed in Airflow 3.0
   |
50 | from airflow.www.utils import get_sensitive_variables_fields, should_hide_value_for_key
51 | 
52 | PY36, PY37, PY38, PY39, PY310, PY311, PY312
   |       ^^^^ AIR302
53 | 
54 | AWSAthenaHook
   |
   = help: Use `sys.version_info` instead

AIR302_names.py:52:13: AIR302 `airflow.PY38` is removed in Airflow 3.0
   |
50 | from airflow.www.utils import get_sensitive_variables_fields, should_hide_value_for_key
51 | 
52 | PY36, PY37, PY38, PY39, PY310, PY311, PY312
   |             ^^^^ AIR302
53 | 
54 | AWSAthenaHook
   |
   = help: Use `sys.version_info` instead

AIR302_names.py:52:19: AIR302 `airflow.PY39` is removed in Airflow 3.0
   |
50 | from airflow.www.utils import get_sensitive_variables_fields, should_hide_value_for_key
51 | 
52 | PY36, PY37, PY38, PY39, PY310, PY311, PY312
   |                   ^^^^ AIR302
53 | 
54 | AWSAthenaHook
   |
   = help: Use `sys.version_info` instead

AIR302_names.py:52:25: AIR302 `airflow.PY310` is removed in Airflow 3.0
   |
50 | from airflow.www.utils import get_sensitive_variables_fields, should_hide_value_for_key
51 | 
52 | PY36, PY37, PY38, PY39, PY310, PY311, PY312
   |                         ^^^^^ AIR302
53 | 
54 | AWSAthenaHook
   |
   = help: Use `sys.version_info` instead

AIR302_names.py:52:32: AIR302 `airflow.PY311` is removed in Airflow 3.0
   |
50 | from airflow.www.utils import get_sensitive_variables_fields, should_hide_value_for_key
51 | 
52 | PY36, PY37, PY38, PY39, PY310, PY311, PY312
   |                                ^^^^^ AIR302
53 | 
54 | AWSAthenaHook
   |
   = help: Use `sys.version_info` instead

AIR302_names.py:52:39: AIR302 `airflow.PY312` is removed in Airflow 3.0
   |
50 | from airflow.www.utils import get_sensitive_variables_fields, should_hide_value_for_key
51 | 
52 | PY36, PY37, PY38, PY39, PY310, PY311, PY312
   |                                       ^^^^^ AIR302
53 | 
54 | AWSAthenaHook
   |
   = help: Use `sys.version_info` instead

AIR302_names.py:54:1: AIR302 `airflow.contrib.aws_athena_hook.AWSAthenaHook` is removed in Airflow 3.0
   |
52 | PY36, PY37, PY38, PY39, PY310, PY311, PY312
53 | 
54 | AWSAthenaHook
   | ^^^^^^^^^^^^^ AIR302
55 | TaskStateTrigger
   |

AIR302_names.py:55:1: AIR302 `airflow.triggers.external_task.TaskStateTrigger` is removed in Airflow 3.0
   |
54 | AWSAthenaHook
55 | TaskStateTrigger
   | ^^^^^^^^^^^^^^^^ AIR302
56 | 
57 | requires_access
   |

AIR302_names.py:57:1: AIR302 `airflow.api_connexion.security.requires_access` is removed in Airflow 3.0
   |
55 | TaskStateTrigger
56 | 
57 | requires_access
   | ^^^^^^^^^^^^^^^ AIR302
58 | 
59 | AllowListValidator
   |
   = help: Use `airflow.api_connexion.security.requires_access_*` instead

AIR302_names.py:59:1: AIR302 `airflow.metrics.validators.AllowListValidator` is removed in Airflow 3.0
   |
57 | requires_access
58 | 
59 | AllowListValidator
   | ^^^^^^^^^^^^^^^^^^ AIR302
60 | BlockListValidator
   |
   = help: Use `airflow.metrics.validators.PatternAllowListValidator` instead

AIR302_names.py:60:1: AIR302 `airflow.metrics.validators.BlockListValidator` is removed in Airflow 3.0
   |
59 | AllowListValidator
60 | BlockListValidator
   | ^^^^^^^^^^^^^^^^^^ AIR302
61 | 
62 | SubDagOperator
   |
   = help: Use `airflow.metrics.validators.PatternBlockListValidator` instead

AIR302_names.py:62:1: AIR302 `airflow.operators.subdag.SubDagOperator` is removed in Airflow 3.0
   |
60 | BlockListValidator
61 | 
62 | SubDagOperator
   | ^^^^^^^^^^^^^^ AIR302
63 | 
64 | dates.date_range
   |

AIR302_names.py:64:7: AIR302 `airflow.utils.dates.date_range` is removed in Airflow 3.0
   |
62 | SubDagOperator
63 | 
64 | dates.date_range
   |       ^^^^^^^^^^ AIR302
65 | dates.days_ago
   |
   = help: Use `airflow.timetables.` instead

AIR302_names.py:65:7: AIR302 `airflow.utils.dates.days_ago` is removed in Airflow 3.0
   |
64 | dates.date_range
65 | dates.days_ago
   |       ^^^^^^^^ AIR302
66 | 
67 | date_range
   |
   = help: Use `pendulum.today('UTC').add(days=-N, ...)` instead

AIR302_names.py:67:1: AIR302 `airflow.utils.dates.date_range` is removed in Airflow 3.0
   |
65 | dates.days_ago
66 | 
67 | date_range
   | ^^^^^^^^^^ AIR302
68 | days_ago
69 | parse_execution_date
   |
   = help: Use `airflow.timetables.` instead

AIR302_names.py:68:1: AIR302 `airflow.utils.dates.days_ago` is removed in Airflow 3.0
   |
67 | date_range
68 | days_ago
   | ^^^^^^^^ AIR302
69 | parse_execution_date
70 | round_time
   |
   = help: Use `pendulum.today('UTC').add(days=-N, ...)` instead

AIR302_names.py:69:1: AIR302 `airflow.utils.dates.parse_execution_date` is removed in Airflow 3.0
   |
67 | date_range
68 | days_ago
69 | parse_execution_date
   | ^^^^^^^^^^^^^^^^^^^^ AIR302
70 | round_time
71 | scale_time_units
   |

AIR302_names.py:70:1: AIR302 `airflow.utils.dates.round_time` is removed in Airflow 3.0
   |
68 | days_ago
69 | parse_execution_date
70 | round_time
   | ^^^^^^^^^^ AIR302
71 | scale_time_units
72 | infer_time_unit
   |

AIR302_names.py:71:1: AIR302 `airflow.utils.dates.scale_time_units` is removed in Airflow 3.0
   |
69 | parse_execution_date
70 | round_time
71 | scale_time_units
   | ^^^^^^^^^^^^^^^^ AIR302
72 | infer_time_unit
   |

AIR302_names.py:72:1: AIR302 `airflow.utils.dates.infer_time_unit` is removed in Airflow 3.0
   |
70 | round_time
71 | scale_time_units
72 | infer_time_unit
   | ^^^^^^^^^^^^^^^ AIR302
   |

AIR302_names.py:79:1: AIR302 `airflow.configuration.get` is removed in Airflow 3.0
   |
77 | dates.datetime_to_nano
78 | 
79 | get, getboolean, getfloat, getint, has_option, remove_option, as_dict, set
   | ^^^ AIR302
80 | 
81 | get_connection, load_connections
   |
   = help: Use `airflow.configuration.conf.get` instead

AIR302_names.py:79:6: AIR302 `airflow.configuration.getboolean` is removed in Airflow 3.0
   |
77 | dates.datetime_to_nano
78 | 
79 | get, getboolean, getfloat, getint, has_option, remove_option, as_dict, set
   |      ^^^^^^^^^^ AIR302
80 | 
81 | get_connection, load_connections
   |
   = help: Use `airflow.configuration.conf.getboolean` instead

AIR302_names.py:79:18: AIR302 `airflow.configuration.getfloat` is removed in Airflow 3.0
   |
77 | dates.datetime_to_nano
78 | 
79 | get, getboolean, getfloat, getint, has_option, remove_option, as_dict, set
   |                  ^^^^^^^^ AIR302
80 | 
81 | get_connection, load_connections
   |
   = help: Use `airflow.configuration.conf.getfloat` instead

AIR302_names.py:79:28: AIR302 `airflow.configuration.getint` is removed in Airflow 3.0
   |
77 | dates.datetime_to_nano
78 | 
79 | get, getboolean, getfloat, getint, has_option, remove_option, as_dict, set
   |                            ^^^^^^ AIR302
80 | 
81 | get_connection, load_connections
   |
   = help: Use `airflow.configuration.conf.getint` instead

AIR302_names.py:79:36: AIR302 `airflow.configuration.has_option` is removed in Airflow 3.0
   |
77 | dates.datetime_to_nano
78 | 
79 | get, getboolean, getfloat, getint, has_option, remove_option, as_dict, set
   |                                    ^^^^^^^^^^ AIR302
80 | 
81 | get_connection, load_connections
   |
   = help: Use `airflow.configuration.conf.has_option` instead

AIR302_names.py:79:48: AIR302 `airflow.configuration.remove_option` is removed in Airflow 3.0
   |
77 | dates.datetime_to_nano
78 | 
79 | get, getboolean, getfloat, getint, has_option, remove_option, as_dict, set
   |                                                ^^^^^^^^^^^^^ AIR302
80 | 
81 | get_connection, load_connections
   |
   = help: Use `airflow.configuration.conf.remove_option` instead

AIR302_names.py:79:63: AIR302 `airflow.configuration.as_dict` is removed in Airflow 3.0
   |
77 | dates.datetime_to_nano
78 | 
79 | get, getboolean, getfloat, getint, has_option, remove_option, as_dict, set
   |                                                               ^^^^^^^ AIR302
80 | 
81 | get_connection, load_connections
   |
   = help: Use `airflow.configuration.conf.as_dict` instead

AIR302_names.py:79:72: AIR302 `airflow.configuration.set` is removed in Airflow 3.0
   |
77 | dates.datetime_to_nano
78 | 
79 | get, getboolean, getfloat, getint, has_option, remove_option, as_dict, set
   |                                                                        ^^^ AIR302
80 | 
81 | get_connection, load_connections
   |
   = help: Use `airflow.configuration.conf.set` instead

AIR302_names.py:81:1: AIR302 `airflow.secrets.local_filesystem.get_connection` is removed in Airflow 3.0
   |
79 | get, getboolean, getfloat, getint, has_option, remove_option, as_dict, set
80 | 
81 | get_connection, load_connections
   | ^^^^^^^^^^^^^^ AIR302
   |
   = help: Use `airflow.secrets.local_filesystem.load_connections_dict` instead

AIR302_names.py:81:17: AIR302 `airflow.secrets.local_filesystem.load_connections` is removed in Airflow 3.0
   |
79 | get, getboolean, getfloat, getint, has_option, remove_option, as_dict, set
80 | 
81 | get_connection, load_connections
   |                 ^^^^^^^^^^^^^^^^ AIR302
   |
   = help: Use `airflow.secrets.local_filesystem.load_connections_dict` instead

AIR302_names.py:84:1: AIR302 `airflow.sensors.external_task_sensor.ExternalTaskSensorLink` is removed in Airflow 3.0
   |
84 | ExternalTaskSensorLink
   | ^^^^^^^^^^^^^^^^^^^^^^ AIR302
85 | BashOperator
86 | BaseBranchOperator
   |
   = help: Use `airflow.sensors.external_task.ExternalTaskSensorLink` instead

AIR302_names.py:85:1: AIR302 `airflow.operators.bash_operator.BashOperator` is removed in Airflow 3.0
   |
84 | ExternalTaskSensorLink
85 | BashOperator
   | ^^^^^^^^^^^^ AIR302
86 | BaseBranchOperator
87 | EmptyOperator, DummyOperator
   |
   = help: Use `airflow.operators.bash.BashOperator` instead

AIR302_names.py:86:1: AIR302 `airflow.operators.branch_operator.BaseBranchOperator` is removed in Airflow 3.0
   |
84 | ExternalTaskSensorLink
85 | BashOperator
86 | BaseBranchOperator
   | ^^^^^^^^^^^^^^^^^^ AIR302
87 | EmptyOperator, DummyOperator
88 | dummy_operator.EmptyOperator
   |
   = help: Use `airflow.operators.branch.BaseBranchOperator` instead

AIR302_names.py:87:16: AIR302 `airflow.operators.dummy.DummyOperator` is removed in Airflow 3.0
   |
85 | BashOperator
86 | BaseBranchOperator
87 | EmptyOperator, DummyOperator
   |                ^^^^^^^^^^^^^ AIR302
88 | dummy_operator.EmptyOperator
89 | dummy_operator.DummyOperator
   |
   = help: Use `airflow.operators.empty.EmptyOperator` instead

AIR302_names.py:88:16: AIR302 `airflow.operators.dummy_operator.EmptyOperator` is removed in Airflow 3.0
   |
86 | BaseBranchOperator
87 | EmptyOperator, DummyOperator
88 | dummy_operator.EmptyOperator
   |                ^^^^^^^^^^^^^ AIR302
89 | dummy_operator.DummyOperator
90 | EmailOperator
   |
   = help: Use `airflow.operators.empty.EmptyOperator` instead

AIR302_names.py:89:16: AIR302 `airflow.operators.dummy_operator.DummyOperator` is removed in Airflow 3.0
   |
87 | EmptyOperator, DummyOperator
88 | dummy_operator.EmptyOperator
89 | dummy_operator.DummyOperator
   |                ^^^^^^^^^^^^^ AIR302
90 | EmailOperator
91 | BaseSensorOperator
   |
   = help: Use `airflow.operators.empty.EmptyOperator` instead

AIR302_names.py:90:1: AIR302 `airflow.operators.email_operator.EmailOperator` is removed in Airflow 3.0
   |
88 | dummy_operator.EmptyOperator
89 | dummy_operator.DummyOperator
90 | EmailOperator
   | ^^^^^^^^^^^^^ AIR302
91 | BaseSensorOperator
92 | DateTimeSensor
   |
   = help: Use `airflow.operators.email.EmailOperator` instead

AIR302_names.py:91:1: AIR302 `airflow.sensors.base_sensor_operator.BaseSensorOperator` is removed in Airflow 3.0
   |
89 | dummy_operator.DummyOperator
90 | EmailOperator
91 | BaseSensorOperator
   | ^^^^^^^^^^^^^^^^^^ AIR302
92 | DateTimeSensor
93 | (ExternalTaskMarker, ExternalTaskSensor, ExternalTaskSensorLink)
   |
   = help: Use `airflow.sensors.base.BaseSensorOperator` instead

AIR302_names.py:92:1: AIR302 `airflow.sensors.date_time_sensor.DateTimeSensor` is removed in Airflow 3.0
   |
90 | EmailOperator
91 | BaseSensorOperator
92 | DateTimeSensor
   | ^^^^^^^^^^^^^^ AIR302
93 | (ExternalTaskMarker, ExternalTaskSensor, ExternalTaskSensorLink)
94 | TimeDeltaSensor
   |
   = help: Use `airflow.sensors.date_time.DateTimeSensor` instead

AIR302_names.py:93:2: AIR302 `airflow.sensors.external_task_sensor.ExternalTaskMarker` is removed in Airflow 3.0
   |
91 | BaseSensorOperator
92 | DateTimeSensor
93 | (ExternalTaskMarker, ExternalTaskSensor, ExternalTaskSensorLink)
   |  ^^^^^^^^^^^^^^^^^^ AIR302
94 | TimeDeltaSensor
   |
   = help: Use `airflow.sensors.external_task.ExternalTaskMarker` instead

AIR302_names.py:93:22: AIR302 `airflow.sensors.external_task_sensor.ExternalTaskSensor` is removed in Airflow 3.0
   |
91 | BaseSensorOperator
92 | DateTimeSensor
93 | (ExternalTaskMarker, ExternalTaskSensor, ExternalTaskSensorLink)
   |                      ^^^^^^^^^^^^^^^^^^ AIR302
94 | TimeDeltaSensor
   |
   = help: Use `airflow.sensors.external_task.ExternalTaskSensor` instead

AIR302_names.py:93:42: AIR302 `airflow.sensors.external_task_sensor.ExternalTaskSensorLink` is removed in Airflow 3.0
   |
91 | BaseSensorOperator
92 | DateTimeSensor
93 | (ExternalTaskMarker, ExternalTaskSensor, ExternalTaskSensorLink)
   |                                          ^^^^^^^^^^^^^^^^^^^^^^ AIR302
94 | TimeDeltaSensor
   |
   = help: Use `airflow.sensors.external_task.ExternalTaskSensorLink` instead

AIR302_names.py:94:1: AIR302 `airflow.sensors.time_delta_sensor.TimeDeltaSensor` is removed in Airflow 3.0
   |
92 | DateTimeSensor
93 | (ExternalTaskMarker, ExternalTaskSensor, ExternalTaskSensorLink)
94 | TimeDeltaSensor
   | ^^^^^^^^^^^^^^^ AIR302
95 | 
96 | apply_defaults
   |
   = help: Use `airflow.sensors.time_delta.TimeDeltaSensor` instead

AIR302_names.py:96:1: AIR302 `airflow.utils.decorators.apply_defaults` is removed in Airflow 3.0; `apply_defaults` is now unconditionally done and can be safely removed.
   |
94 | TimeDeltaSensor
95 | 
96 | apply_defaults
   | ^^^^^^^^^^^^^^ AIR302
97 | 
98 | TemporaryDirectory
   |

AIR302_names.py:98:1: AIR302 `airflow.utils.file.TemporaryDirectory` is removed in Airflow 3.0
   |
96 | apply_defaults
97 | 
98 | TemporaryDirectory
   | ^^^^^^^^^^^^^^^^^^ AIR302
99 | mkdirs
   |

AIR302_names.py:99:1: AIR302 `airflow.utils.file.mkdirs` is removed in Airflow 3.0
    |
 98 | TemporaryDirectory
 99 | mkdirs
    | ^^^^^^ AIR302
100 | 
101 | chain
    |
    = help: Use `pendulum.today('UTC').add(days=-N, ...)` instead

AIR302_names.py:101:1: AIR302 `airflow.utils.helpers.chain` is removed in Airflow 3.0
    |
 99 | mkdirs
100 | 
101 | chain
    | ^^^^^ AIR302
102 | cross_downstream
    |
    = help: Use `airflow.models.baseoperator.chain` instead

AIR302_names.py:102:1: AIR302 `airflow.utils.helpers.cross_downstream` is removed in Airflow 3.0
    |
101 | chain
102 | cross_downstream
    | ^^^^^^^^^^^^^^^^ AIR302
103 | 
104 | SHUTDOWN
    |
    = help: Use `airflow.models.baseoperator.cross_downstream` instead

AIR302_names.py:104:1: AIR302 `airflow.utils.state.SHUTDOWN` is removed in Airflow 3.0
    |
102 | cross_downstream
103 | 
104 | SHUTDOWN
    | ^^^^^^^^ AIR302
105 | terminating_states
    |

AIR302_names.py:105:1: AIR302 `airflow.utils.state.terminating_states` is removed in Airflow 3.0
    |
104 | SHUTDOWN
105 | terminating_states
    | ^^^^^^^^^^^^^^^^^^ AIR302
106 | 
107 | TriggerRule.DUMMY
    |

AIR302_names.py:107:13: AIR302 `airflow.utils.trigger_rule.TriggerRule.DUMMY` is removed in Airflow 3.0
    |
105 | terminating_states
106 | 
107 | TriggerRule.DUMMY
    |             ^^^^^ AIR302
108 | TriggerRule.NONE_FAILED_OR_SKIPPED
    |

AIR302_names.py:108:13: AIR302 `airflow.utils.trigger_rule.TriggerRule.NONE_FAILED_OR_SKIPPED` is removed in Airflow 3.0
    |
107 | TriggerRule.DUMMY
108 | TriggerRule.NONE_FAILED_OR_SKIPPED
    |             ^^^^^^^^^^^^^^^^^^^^^^ AIR302
109 | 
110 | test_cycle
    |

AIR302_names.py:110:1: AIR302 `airflow.utils.dag_cycle_tester.test_cycle` is removed in Airflow 3.0
    |
108 | TriggerRule.NONE_FAILED_OR_SKIPPED
109 | 
110 | test_cycle
    | ^^^^^^^^^^ AIR302
111 | 
112 | has_access
    |

AIR302_names.py:112:1: AIR302 `airflow.www.auth.has_access` is removed in Airflow 3.0
    |
110 | test_cycle
111 | 
112 | has_access
    | ^^^^^^^^^^ AIR302
113 | get_sensitive_variables_fields, should_hide_value_for_key
    |
    = help: Use `airflow.www.auth.has_access_*` instead

AIR302_names.py:113:1: AIR302 `airflow.www.utils.get_sensitive_variables_fields` is removed in Airflow 3.0
    |
112 | has_access
113 | get_sensitive_variables_fields, should_hide_value_for_key
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ AIR302
    |
    = help: Use `airflow.utils.log.secrets_masker.get_sensitive_variables_fields` instead

AIR302_names.py:113:33: AIR302 `airflow.www.utils.should_hide_value_for_key` is removed in Airflow 3.0
    |
112 | has_access
113 | get_sensitive_variables_fields, should_hide_value_for_key
    |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^ AIR302
    |
    = help: Use `airflow.utils.log.secrets_masker.should_hide_value_for_key` instead
